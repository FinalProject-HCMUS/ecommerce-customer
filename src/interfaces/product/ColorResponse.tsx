export interface ColorResponse {
  /**
   * Unique identifier of the color
   * @example "550e8400-e29b-41d4-a716-446655440000"
   */
  id: string;

  /**
   * Name of the color
   * @example "Red"
   */
  name: string;

  /**
   * Date and time when the color was created
   */
  createdAt: string;

  /**
   * User who created the color
   * @example "admin"
   */
  createdBy: string;

  /**
   * Date and time when the color was last updated
   */
  updatedAt: string;

  /**
   * User who last updated the color
   * @example "admin"
   */
  updatedBy: string;
}
