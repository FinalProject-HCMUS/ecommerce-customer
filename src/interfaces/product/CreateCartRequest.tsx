export interface CreateCartItemRequest {
  quantity: number;
  userId: string;
  itemId: string;
}
