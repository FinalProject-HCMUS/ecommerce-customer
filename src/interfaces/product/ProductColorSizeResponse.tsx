export interface ProductColorSizeResponse {
  /**
   * Quantity of the product_color_size
   * @example 10
   */
  quantity: number;

  /**
   * ID of the product
   * @example "123e4567-e89b-12d3-a456-426614174000"
   */
  productId: string;

  /**
   * ID of the color
   * @example "456e7890-b12c-34d5-e678-123456789abc"
   */
  colorId: string;

  /**
   * ID of the size
   * @example "789e0123-c45d-67f8-9012-3456789abcde"
   */
  sizeId: string;

  /**
   * Date and time when the product_color_size was created
   */
  createdAt: string;

  /**
   * User who created the product_color_size
   * @example "admin"
   */
  createdBy: string;

  /**
   * Date and time when the product_color_size was last updated
   */
  updatedAt: string;

  /**
   * User who last updated the product_color_size
   * @example "admin"
   */
  updatedBy: string;
}
