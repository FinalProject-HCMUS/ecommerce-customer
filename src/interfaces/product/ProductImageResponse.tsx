export interface ProductImageResponse {
  id: string;
  url: string;
  productId: string;
  createdAt: string;
  createdBy: string;
  updatedAt: string;
  updatedBy: string;
}
