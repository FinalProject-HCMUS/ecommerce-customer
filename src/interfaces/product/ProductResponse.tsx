export interface ProductResponse {
  /**
   * Unique identifier of the product
   * @example "550e8400-e29b-41d4-a716-446655440000"
   */
  id: string;

  /**
   * Name of the product
   * @example "iPhone 12"
   */
  name: string;

  /**
   * Description of the product
   * @example "The latest iPhone model"
   */
  description: string;

  /**
   * Cost of the product
   * @example 800
   */
  cost: number;

  /**
   * Total quantity of the product
   * @example 100
   */
  total: number;

  /**
   * Price of the product
   * @example 1000
   */
  price: number;

  /**
   * Discount percentage of the product
   * @example 0.1 // 10%
   */
  discountPercent: number;

  /**
   * Whether the product is enabled
   * @example true
   */
  enable: boolean;

  /**
   * Whether the product is in stock
   * @example true
   */
  inStock: boolean;

  /**
   * URL of the main image of the product
   * @example "https://example.com/image.jpg"
   */
  mainImageUrl: string;

  /**
   * Average rating of the product
   * @example 4.5
   */
  averageRating: number;

  /**
   * Number of reviews of the product
   * @example 100
   */
  reviewCount: number;

  /**
   * ID of the category this product belongs to
   * @example "550e8400-e29b-41d4-a716-446655440000"
   */
  categoryId: string;

  /**
   * Date and time when the product was created
   */
  createdAt: string;

  /**
   * Date and time when the product was last updated
   */
  updatedAt: string;

  /**
   * User who created the product
   * @example "admin"
   */
  createdBy: string;

  /**
   * User who last updated the product
   * @example "admin"
   */
  updatedBy: string;
}
