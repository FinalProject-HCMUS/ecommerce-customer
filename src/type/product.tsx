export interface Product {
  id: number;
  name: string;
  title: string;
  price: number;
  originalPrice?: number;
  rating: number;
  category: string;
  color: string;
  image: string;
  sizes: string[];
  reviews: number;
}
