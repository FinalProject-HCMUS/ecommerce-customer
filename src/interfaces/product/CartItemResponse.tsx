import { ProductResponse } from './ProductResponse';

export interface CartItemResponse {
  id: string;
  quantity: number;
  product: ProductResponse;
  color: string;
  size: string;
  userId: string;
  itemId: string;
}
