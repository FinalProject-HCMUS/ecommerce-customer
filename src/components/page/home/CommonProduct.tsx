import React, { useState } from 'react';
import ProductCard from '../../shared/product-card/ProductCard';
import { Product } from '../../../type/product';

interface CommonProductsProps {
  title: string;
  data: Product[];
}

const CommonProducts: React.FC<CommonProductsProps> = ({ title, data }) => {
  const [visibleCount, setVisibleCount] = useState(5); // Number of products to show initially

  const handleSeeMore = () => {
    setVisibleCount((prevCount) => prevCount + 5); // Load 10 more products
  };

  return (
    <section className="py-16 bg-gray-100">
      <div className="container mx-auto px-4">
        <h2 className="text-3xl font-bold text-center mb-12">{title}</h2>
        <div className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-5 gap-2">
          {data.slice(0, visibleCount).map((product) => (
            <ProductCard key={product.id} product={product} />
          ))}
        </div>

        {visibleCount < data.length && ( // Show "See more" only if there are more products to load
          <div className="text-center mt-12">
            <button
              className="btn-primary border-2 border-gray-300 px-8 py-2 rounded-full hover:bg-gray-100 transition-colors"
              onClick={handleSeeMore}
            >
              See more
            </button>
          </div>
        )}
      </div>
    </section>
  );
};

export default CommonProducts;