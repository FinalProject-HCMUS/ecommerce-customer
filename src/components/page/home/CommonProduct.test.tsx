import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import CommonProduct from './CommonProduct';
import { ProductResponse } from '../../../interfaces/product/ProductResponse';

// Mock the constants - important to set VISIBLE_PRODUCT for testing
jest.mock('../../../constants', () => ({
  common: {
    VISIBLE_PRODUCT: 2  // Set to 2 for easier testing of pagination
  }
}));

// Mock the ProductCard component
jest.mock('../../shared/product-card/ProductCard', () => {
  return {
    __esModule: true,
    default: ({ product }: { product: ProductResponse }) => (
      <div data-testid={`product-card-${product.id}`}>
        <div className="product-name">{product.name}</div>
        <div className="product-price">{product.price}</div>
      </div>
    ),
  };
});

// Mock the translation function
jest.mock('../../../helpers/i18n', () => ({
  t: jest.fn((key) => key),
}));

describe('CommonProduct', () => {
  // Mock product data
  const mockProducts: ProductResponse[] = [
    {
      id: '1',
      name: 'Test Product 1',
      description: 'Test description 1',
      cost: 80000,
      total: 100,
      price: 100000,
      discountPercent: 10,
      enable: true,
      inStock: true,
      mainImageUrl: '/test-image1.jpg',
      averageRating: 4.5,
      reviewCount: 12,
      categoryId: 'category-1',
      createdAt: '2023-01-01',
      updatedAt: '2023-01-01',
      createdBy: 'admin',
      updatedBy: 'admin',
    },
    {
      id: '2',
      name: 'Test Product 2',
      description: 'Test description 2',
      cost: 90000,
      total: 50,
      price: 120000,
      discountPercent: 5,
      enable: true,
      inStock: true,
      mainImageUrl: '/test-image2.jpg',
      averageRating: 4.0,
      reviewCount: 8,
      categoryId: 'category-1',
      createdAt: '2023-01-01',
      updatedAt: '2023-01-01',
      createdBy: 'admin',
      updatedBy: 'admin',
    },
    {
      id: '3',
      name: 'Test Product 3',
      description: 'Test description 3',
      cost: 100000,
      total: 75,
      price: 150000,
      discountPercent: 15,
      enable: true,
      inStock: true,
      mainImageUrl: '/test-image3.jpg',
      averageRating: 5.0,
      reviewCount: 20,
      categoryId: 'category-2',
      createdAt: '2023-01-01',
      updatedAt: '2023-01-01',
      createdBy: 'admin',
      updatedBy: 'admin',
    },
    {
      id: '4',
      name: 'Test Product 4',
      description: 'Test description 4',
      cost: 110000,
      total: 85,
      price: 160000,
      discountPercent: 20,
      enable: true,
      inStock: true,
      mainImageUrl: '/test-image4.jpg',
      averageRating: 4.2,
      reviewCount: 15,
      categoryId: 'category-2',
      createdAt: '2023-01-01',
      updatedAt: '2023-01-01',
      createdBy: 'admin',
      updatedBy: 'admin',
    },
  ];

  const renderWithRouter = (ui: React.ReactElement) => {
    return render(ui, { wrapper: BrowserRouter });
  };

  test('renders the component with title', () => {
    renderWithRouter(<CommonProduct title="Featured Products" data={mockProducts} />);
    
    // Check if the title is rendered
    expect(screen.getByText('Featured Products')).toBeInTheDocument();
  });

  test('initially renders only VISIBLE_PRODUCT number of products', () => {
    renderWithRouter(<CommonProduct title="Featured Products" data={mockProducts} />);
    
    // Should render only first 2 products initially (VISIBLE_PRODUCT = 2)
    expect(screen.getByTestId('product-card-1')).toBeInTheDocument();
    expect(screen.getByTestId('product-card-2')).toBeInTheDocument();
    expect(screen.queryByTestId('product-card-3')).not.toBeInTheDocument();
    expect(screen.queryByTestId('product-card-4')).not.toBeInTheDocument();
  });

  test('shows "See More" button when there are more products to show', () => {
    renderWithRouter(<CommonProduct title="Featured Products" data={mockProducts} />);
    
    // See More button should be visible since we have 4 products and VISIBLE_PRODUCT = 2
    expect(screen.getByText('btn.seeMore')).toBeInTheDocument();
  });

  test('clicking "See More" loads more products', () => {
    renderWithRouter(<CommonProduct title="Featured Products" data={mockProducts} />);
    
    // Initially should not see product 3 and 4
    expect(screen.queryByTestId('product-card-3')).not.toBeInTheDocument();
    expect(screen.queryByTestId('product-card-4')).not.toBeInTheDocument();
    
    // Click See More button
    fireEvent.click(screen.getByText('btn.seeMore'));
    
    // Now should see all 4 products
    expect(screen.getByTestId('product-card-1')).toBeInTheDocument();
    expect(screen.getByTestId('product-card-2')).toBeInTheDocument();
    expect(screen.getByTestId('product-card-3')).toBeInTheDocument();
    expect(screen.getByTestId('product-card-4')).toBeInTheDocument();
    
    // Should now show "See Less" button
    expect(screen.getByText('btn.seeLess')).toBeInTheDocument();
  });

  test('clicking "See Less" reduces the number of products shown', () => {
    renderWithRouter(<CommonProduct title="Featured Products" data={mockProducts} />);
    
    // Click See More to show all products
    fireEvent.click(screen.getByText('btn.seeMore'));
    
    // Verify all products are shown
    expect(screen.getByTestId('product-card-3')).toBeInTheDocument();
    expect(screen.getByTestId('product-card-4')).toBeInTheDocument();
    
    // Click See Less
    fireEvent.click(screen.getByText('btn.seeLess'));
    
    // Should go back to showing only 2 products
    expect(screen.getByTestId('product-card-1')).toBeInTheDocument();
    expect(screen.getByTestId('product-card-2')).toBeInTheDocument();
    expect(screen.queryByTestId('product-card-3')).not.toBeInTheDocument();
    expect(screen.queryByTestId('product-card-4')).not.toBeInTheDocument();
  });

  test('does not show "See More" button when all products can be shown initially', () => {
    renderWithRouter(<CommonProduct title="Featured Products" data={mockProducts.slice(0, 2)} />);
    
    // See More button should not be visible since we only have 2 products and VISIBLE_PRODUCT = 2
    expect(screen.queryByText('btn.seeMore')).not.toBeInTheDocument();
  });

  test('renders empty grid when no products are provided', () => {
    renderWithRouter(<CommonProduct title="Featured Products" data={[]} />);
    
    // Should render the section but with no product cards
    expect(screen.getByText('Featured Products')).toBeInTheDocument();
    expect(screen.queryByTestId(/product-card-/)).not.toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { container } = renderWithRouter(<CommonProduct title="Featured Products" data={mockProducts} />);
    expect(container).toMatchSnapshot();
  });
});