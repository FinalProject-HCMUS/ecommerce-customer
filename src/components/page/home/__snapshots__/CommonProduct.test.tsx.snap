// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonProduct matches snapshot 1`] = `
<div>
  <section
    class="py-16 bg-gray-100"
  >
    <div
      class="px-8"
    >
      <h2
        class="text-3xl font-bold text-center mb-12"
      >
        Featured Products
      </h2>
      <div
        class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-6 gap-2"
      >
        <div
          data-testid="product-card-1"
        >
          <div
            class="product-name"
          >
            Test Product 1
          </div>
          <div
            class="product-price"
          >
            100000
          </div>
        </div>
        <div
          data-testid="product-card-2"
        >
          <div
            class="product-name"
          >
            Test Product 2
          </div>
          <div
            class="product-price"
          >
            120000
          </div>
        </div>
      </div>
      <div
        class="text-center mt-12"
      >
        <button
          class="btn-primary border-2 border-gray-300 px-8 py-2 rounded-full hover:bg-gray-100 transition-colors mr-4"
        >
          btn.seeMore
        </button>
      </div>
    </div>
  </section>
</div>
`;
